app/scripts/controllers/transactions/index.js

```diff
@@ -781,18 +776,30 @@ export default class TransactionController extends EventEmitter {
    *  fragment for
    * @param {valueOf<TransactionMetaMetricsEvent>} event - event type to create
    * @param {string} actionId - actionId passed from UI
+   * @param {boolean} isFragmentExists - whether fragment exist or not
+   * @param {object} updateParams - Update payload sent from UI
    */
-  async createTransactionEventFragment(transactionId, event, actionId) {
+  async createOrUpdateTransactionEventFragment(
+    transactionId,
+    event,
+    actionId,
+    isFragmentExists,
+    updateParams,
+  ) {
     const txMeta = this.txStateManager.getTransaction(transactionId);
     const { properties, sensitiveProperties } =
       await this._buildEventFragmentProperties(txMeta);
-    this._createTransactionEventFragment(
-      txMeta,
+    this.emit('transaction-metric', {
```

Comment:
Rather than coupling all these new events to the concept of metrics, could we instead decouple it entirely and just announce domain things that are happening like `transaction-added`, `transaction-dropped`, `transaction-failed` etc?

It may simplify the controller alot as it would no longer be concerned with metrics at all but rather just good at communicating what's happening to anyone that is listening.

It would then be the responsibility of the listeners to generate the metrics properties and worry about which fragments exist etc.

Reply:
Sure, as I mentioned in the comment I made below, this was my first intention but had a fear of moving out a really big piece `_buildEventFragmentProperties` from controller as it was collecting lot of domain specific info.